### PR TITLE
Develop book controller

### PR DESCRIPTION
**Pull request by petro2061**
Added web controller layer for Book model. The DTO approach for obtaining and displaying entity properties has also been implemented. A library was used for this Mapstruct.

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -67,6 +69,12 @@
 			<artifactId>h2</artifactId>
 			<scope>test</scope>
 		</dependency>
+
+		<dependency>
+			<groupId>org.mapstruct</groupId>
+			<artifactId>mapstruct</artifactId>
+			<version>1.5.5.Final</version>
```
Comment (Elena-Bruyako) [inline pom.xml]: use properties for mapstruct version here and in plugin section also

File: src/main/java/project/onlinebookstore/controller/BookController.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+import project.onlinebookstore.dto.BookDto;
+import project.onlinebookstore.dto.CreateBookRequestDto;
+import project.onlinebookstore.service.BookService;
+
+@RequiredArgsConstructor
+@RestController
+@RequestMapping("/api/books")
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/controller/BookController.java]:
```suggestion
@RequestMapping("/books")
```

Add `server.servlet.context-path=/api` to application.properties

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job! Left comments

**Comment by petro2061** [comment]
Added new properies to pom.xml file and application.properties. Also fixed request maping annotation.

**Review by Elena-Bruyako — APPROVED** [review]
Well done!

**Comment by petro2061** [comment]
Reworked the implementation of the book repository using Spring JPA. Added liquibase to project.

File: src/main/java/project/onlinebookstore/model/Book.java [hunk 3]
```diff
@@ -7,11 +7,19 @@
...
-import lombok.Data;
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+import lombok.Setter;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.Where;
 
+@Setter
+@Getter
+@RequiredArgsConstructor
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/model/Book.java]:
```suggestion
```

File: src/main/java/project/onlinebookstore/model/Book.java [hunk 4]
```diff
@@ -7,11 +7,19 @@
...
+@SQLDelete(sql = "UPDATE books SET is_deleted = true WHERE id =?")
+@Where(clause = "is_deleted = false")
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/model/Book.java]:
```suggestion
@SQLRestriction(value = "is_deleted = false")
```

`@Where` is deprecated

File: src/main/java/project/onlinebookstore/model/Book.java [hunk 5]
```diff
@@ -26,4 +34,20 @@ public class Book {
     private BigDecimal price;
     private String description;
     private String coverImage;
+    @Column(nullable = false)
+    private boolean isDeleted = false;
+
+    @Override
+    public String toString() {
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/model/Book.java]: remove

File: src/main/java/project/onlinebookstore/service/impl/BookServiceImpl.java [hunk 6]
```diff
@@ -1,24 +1,63 @@
...
-    @Autowired
-    private BookRepository bookRepository;
+    private final BookRepository bookRepository;
+    private final BookMapper bookMapper;
 
     @Override
-    public Book save(Book book) {
-        return bookRepository.save(book);
+    public BookDto save(CreateBookRequestDto bookRequestDto) {
+        Book bookModelForSave =
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/service/impl/BookServiceImpl.java]:
```suggestion
        Book book =
```

File: src/main/java/project/onlinebookstore/service/impl/BookServiceImpl.java [hunk 7]
```diff
@@ -1,24 +1,63 @@
...
+                bookRepository.save(bookMapper.toBookModel(bookRequestDto));
+        return bookMapper.toBookDto(bookModelForSave);
     }
 
     @Override
-    public List<Book> findAll() {
-        return bookRepository.findAll();
+    public List<BookDto> findAll() {
+        List<Book> allBookFromRepository = bookRepository.findAll();
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/service/impl/BookServiceImpl.java]:
```suggestion
        return bookRepository.findAll()
                .stream()
                .map(bookMapper::toBookDto)
                .toList();
```

File: src/main/java/project/onlinebookstore/service/impl/BookServiceImpl.java [hunk 8]
```diff
@@ -1,24 +1,63 @@
...
+    @Override
+    public void deleteById(Long id) {
+        bookRepository.deleteById(id);
+    }
+
+    @Override
+    public BookDto updateById(Long id, CreateBookRequestDto updateBookRequestDto) {
+        Book bookByIdForUpdate = bookRepository.findById(id).orElseThrow(() ->
+                new EntityNotFoundException("Can not find entity by id: " + id));
+        bookByIdForUpdate.setId(id);
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/service/impl/BookServiceImpl.java]:
```suggestion
        bookMapper.updateBookFromDto(bookDto, book);
        return bookMapper.toDto(bookRepository.save(book));
```

File: src/main/java/project/onlinebookstore/mapper/BookMapper.java [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+import project.onlinebookstore.config.MapperConfig;
+import project.onlinebookstore.dto.BookDto;
+import project.onlinebookstore.dto.CreateBookRequestDto;
+import project.onlinebookstore.model.Book;
+
+@Mapper(config = MapperConfig.class)
+public interface BookMapper {
+    BookDto toBookDto(Book book);
+
+    Book toBookModel(CreateBookRequestDto bookRequestDto);
```
Comment (Elena-Bruyako) [inline src/main/java/project/onlinebookstore/mapper/BookMapper.java:14]: Add
`void updateBookFromDto(CreateBookRequestDto book, @MappingTarget Book entity);`

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Good job! let's fix some comments

**Comment by petro2061** [comment]
Aded new method for update book entity to mpper class and use it on servise layer. Also fix problem with codestyle.

**Comment by petro2061** [comment]
Added anotation Transactional and fix problem with codestyle

**Review by Elena-Bruyako — APPROVED** [review]
Good job! 
Don't forget to merge approved PRs and implement new HW in branch without previous changes

**Comment by petro2061** [comment]
All comments from the mentor have been corrected.

**Pull request merged**